### PR TITLE
Refactor character creation fixes #105

### LINKED ISSUE #105: Warnings in Character Creation dialog break deployment
**Bug Summary**
The automated deploy to GitHub pages fails due to warnings present on the `master` branch. The warnings are caused by variables being assigned a value but never read in `src/features/dialog-manager/character-creation/reducer.js`.

**Test Case(s)**

---

**Expected behavior**
There should be no warnings on master, and the deployment action should succeed.

**Observed Behaviour**
There are warnings on master, and the deployment action fails.

**Steps To Reproduce**

1. Run deployment action

**Environment**
- Version:  829f10a
- OS: Any
- Browser: Any 

---

**Code Examples**
N/A
**Stack Trace**
```
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
warning file-loader@2.0.0: Invalid bin field for "file-loader".
warning mini-css-extract-plugin@0.5.0: Invalid bin field for "mini-css-extract-plugin".
warning url-loader@1.1.2: Invalid bin field for "url-loader".
info fsevents@1.2.4: The platform "linux" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.11: The platform "linux" is incompatible with this module.
info "fsevents@1.2.11" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > eslint-config-prettier@6.10.0" has unmet peer dependency "eslint@>=3.14.1".
warning " > eslint-plugin-prettier@3.1.2" has unmet peer dependency "eslint@>= 5.0.0".
[4/4] Building fresh packages...
Done in 37.80s.
yarn run v1.22.4
$ react-scripts build
Creating an optimized production build...

Treating warnings as errors because process.env.CI = true.
Most CI servers set it automatically.

Failed to compile.

./src/features/dialog-manager/reducer.js
  Line 69:  'characterName' is assigned a value but never used   no-unused-vars
  Line 69:  'characterRace' is assigned a value but never used   no-unused-vars
  Line 69:  'characterClass' is assigned a value but never used  no-unused-vars


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
##[error]Process completed with exit code 1.
```
**Screenshots**
N/A
**Error Report**
N/A

---

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Matteas-Eden**
GitHub Issue (If applicable): #105 

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Refactoring (no functional changes, no api changes)

## What is the current behavior?

There are warnings for variables which have been set a value but not read:
- `characterName`
- `characterClass`
- `characterRace`

## What is the new behavior?

By allocating the aforementioned variables as default values when confirming character creation, the warnings are removed.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [ ] Added new tests to the test suite for the feature
- [x] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [ ] Checked licensing of Frameworks/Libraries (if applicable) 

## Other information

In addition to the bugfix described above:
- Character creation actions have been grouped into their own subdirectory of the character creation dialog (`dialog-manager/character-creation/actions`) 

Internal Issue (If applicable): Resolves #105

### CODE REVIEW

File: src/features/dialog-manager/reducer.js [hunk 1]
```diff
@@ -165,10 +165,9 @@ const dialogManagerReducer = (state = initialState, { type, payload }) => {
             return {
                 ...state,
                 character: {
-                    ...character,
-                    characterName: payload.characterName,
-                    characterRace: payload.characterRace,
-                    characterClass: payload.characterClass,
+                    characterName: payload.characterName || characterName,
+                    characterRace: payload.characterRace || characterRace,
+                    characterClass: payload.characterClass || characterClass,
```
Comment (Pyxxil) [inline src/features/dialog-manager/reducer.js:170]: What's the reasoning behind putting the '|| <whatever>' here? As far as I'm concerned, if the <whatever> is undefined, null, or something else 'falsey' then that should be a bug in our logic. I'm happy to hear your thoughts though, because if there's a good reason we should consider adding this to the majority of the reducers.

Good catch on the characterName though.

File: src/features/dialog-manager/reducer.js [hunk 2]
```diff
@@ -43,6 +37,12 @@ const initialState = {
         min_intelligence: 0,
         min_charisma: 0,
     },
+    characterCreation: false,
+    character: {
+        characterName: 'Adventurer',
```
Comment (Pyxxil) [inline src/features/dialog-manager/reducer.js:42]: Is this what we want the default name of the player to be? Will we want this to be a default in the input field for this as well?

File: src/features/dialog-manager/reducer.js (hunk 1 above)
Comment (Pyxxil) [inline src/features/dialog-manager/reducer.js]: What's the reasoning behind putting the '|| [whatever]' here? As far as I'm concerned, if the [whatever] is undefined, null, or something else 'falsey' then that should be a bug in our logic. I'm happy to hear your thoughts though, because if there's a good reason we should consider adding this to the majority of the reducers.

Good catch on the characterName though.

File: src/features/dialog-manager/reducer.js (hunk 2 above)
Comment (Pyxxil) [inline src/features/dialog-manager/reducer.js]: Is this what we want the default name of the player to be? Will we want this to be a default in the input field for this as well?

**Review by Pyxxil — CHANGES_REQUESTED** [review]

**Comment by Pyxxil** [comment]
I should probably mention, because I forgot to do so in the Review comment (apologies). Moving the actions was a good idea, so other than the above it looks good to me.

File: src/features/dialog-manager/reducer.js [hunk 3]
```diff
@@ -9,12 +9,6 @@ const initialState = {
     gameOver: false,
     gameStart: true,
     gameInstructions: false,
-    characterCreation: false,
```
Comment (KimberleyEvans-Parker) [inline src/features/dialog-manager/reducer.js:12]: Why is this moved from this section?  I feel that this is an unnecessary move.  
The reason it was there in the first place was for consistency; everywhere you will find characterCreation, it will always be straight after gameInstructions (which I did on purpose) so that it's easier to find and is consistent.  For those reasons, I think moving them to the bottom is unnecessary.

**Review by KimberleyEvans-Parker — CHANGES_REQUESTED** [review]
I'm rather confused by what this PR is trying to acheive.  What is mentioned in the PR appears to be rather different than what is changed in the files.

File: src/features/dialog-manager/reducer.js (hunk 1 above)
Comment (Matteas-Eden) [reply src/features/dialog-manager/reducer.js]: The reasoning is to force the `character[Name|Race|Class]` variables to be read, which removes the warnings. As you've pointed out the second value is only used if there's something 'wrong' with the first value - something that shouldn't happen given our logic. Although I'm tempted to suggest that a stray runtime error could produce something 'falsey' for the first value, I haven't seen this in testing.

Quite simply, it was the most straightforward way I saw to remove the warnings.

File: src/features/dialog-manager/reducer.js (hunk 3 above)
Comment (Matteas-Eden) [reply src/features/dialog-manager/reducer.js:12]: It was moved so as to be consistent with the other additions that were made. It seemed unclean to have an 'Object block' in the middle of the list, as opposed to being grouped with the other 'Object blocks' (like `abilities` and `ability_minimum`). 
As far as ordering it after `gameInstructions`, I'm afraid I can't see the logic in that decision, as the `characterCreation` dialog comes before `gameInstructions` in the menu flow.

File: src/features/dialog-manager/reducer.js (hunk 2 above)
Comment (Matteas-Eden) [reply src/features/dialog-manager/reducer.js]: The default can be changed to whatever is preferred, I merely put that in because it made sense to me. I'm unsure about making it a default in the input field.

File: src/features/dialog-manager/reducer.js (hunk 1 above)
Comment (Pyxxil) [reply src/features/dialog-manager/reducer.js]: Why not just remove line 69?

Comment (KimberleyEvans-Parker) [reply src/features/dialog-manager/reducer.js]: You could have removed the warnings by just removing line 69.

Comment (Matteas-Eden) [reply src/features/dialog-manager/reducer.js]: Upon reviewing my solution, I agree.

**Review by KimberleyEvans-Parker — DISMISSED** [review]

**Review by KimberleyEvans-Parker — APPROVED** [review]

**Review by Pyxxil — APPROVED** [review]
Looks fine.
